File: Algorithms/Graph Theory/Java Implementations/BreadthFirstSearch.java
/** Breadth First Search algorithm implemented by Arpan Pathak 
 *  Time Complexity O(VE) **/
import java.util.*;
class Edge<T,K>
{
	public T v;   K w;
	public Edge(T vertex,K weight) { v=vertex; w=weight; }
}
class GraphAdjacencyList<T,V>
{
	Map<T, List< Edge<T,V>> > G;
	Set<T> vertices; // vertices is set of all vertices
	GraphAdjacencyList()
	{
		G=new HashMap<T, List<Edge<T,V>> >();
		vertices=new HashSet<>();
	}
	void addEdge(T u,T v,V weight)
	{
		if(!G.containsKey(u))
			G.put(u, new LinkedList<Edge<T,V>>() );
		G.get(u).add(0, new Edge<T,V>(v,weight));
		vertices.add(u); vertices.add(v);	
	}
	void BFS(T start_vertex)
	{
		LinkedList<T> queue=new LinkedList<>();
		Map<T,Boolean> visited=new HashMap<>();
		for(T i: vertices) visited.put(i, (i==start_vertex?true:false));
		queue.add(start_vertex);
		while(!queue.isEmpty())
		{
			T explore=queue.removeFirst();
			System.out.print(explore+",");
			if(!G.containsKey(explore)) continue; // to avoid NullPointerException
			for(Edge<T,V> iterator: G.get(explore))
			{
				if(!visited.get(iterator.v)) {
					visited.put(iterator.v,  true); 
					queue.add(iterator.v);
				}
			}
		}
	}
}
public class BreadthFirstSearch {
	public static void main(String[] args) {
		GraphAdjacencyList<String,Integer> g=new GraphAdjacencyList<>();
		g.addEdge("Goa", "Nagpur", 10);
		g.addEdge("Goa", "Kanpur", 15);
		g.addEdge("Kanpur", "Kolkata", 11);
		g.addEdge("Kolkata", "Delhi", 20);
		g.BFS("Goa");
	}
}

File: Algorithms/Graph Theory/C++ Implementations/BreadthFirstSearch.cpp
/** Breadth First Search algorithm implemented by Arpan Pathak
 *  Time Complexity O(VE) **/
#include <bits/stdc++.h>
using namespace std;
/* Edge Class will contain description of edge (end vertex and weight) */
template<class K,class V>
class Edge
{
        public:
            K vertex; V weight;
            Edge(K vertex,V weight): vertex(vertex),weight(weight) { }
};
/** Graph Template class where K specifies data type of vertex value and V specifies data type of edge weight **/
template<class K,class V>
class Graph
{
    map<K,list<Edge<K,V>> > G; // map will map each vertex to a linked list which contains all adjacent vertices
    set<K> Vertices; // set is used to keep track of all vertices because we are using adjacency list

    public:
        void addEdge(K u,K v,V w) { G[u].push_front(Edge<K,V>(v,w));  Vertices.insert(u); Vertices.insert(v); }
        void BFS(K start_vertex)
        {
            queue<K> que; map<K,bool> visited;
            for(auto i: Vertices)
                visited[i]=(i==start_vertex?true:false);
            que.push(start_vertex);
            while(!que.empty())
            {
                K explore=que.front(); que.pop(); cout<<explore<<",";
                for(auto i: G[explore])
                {
                    if(!visited[i.vertex]){
                        que.push(i.vertex); visited[i.vertex]=true;
                    }
                }
            }
        }
};
int main()
{
    Graph<string,int> g;
    g.addEdge("A", "B", 10);
    g.addEdge("A", "C", 15);
    g.addEdge("A", "D", 11);
    g.addEdge("D", "E", 20);
    g.BFS("A");
}
